File: code/654.maximum-binary-tree.java
/*
 * @lc app=leetcode id=654 lang=java
 *
 * [654] Maximum Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        return buildTree(nums, 0, nums.length);
    }

    public TreeNode buildTree(int[] nums, int leftIndex, int rightIndex) {
        if (rightIndex - leftIndex < 1) {
            return null;
        }
        if (rightIndex - leftIndex == 1) {
            return new TreeNode(nums[leftIndex]);
        }
        int maxIndex = leftIndex;
        int maxValue = nums[maxIndex];
        for (int i = leftIndex + 1; i < rightIndex; i++) {
            if (nums[i] > maxValue) {
                maxValue = nums[i];
                maxIndex = i;
            }
        }
        TreeNode root = new TreeNode(maxValue);
        root.left = buildTree(nums, leftIndex, maxIndex);
        root.right = buildTree(nums, maxIndex + 1, rightIndex);
        return root;
    }
}
// @lc code=end

File: code/654.maximum-binary-tree.go
/*
 * @lc app=leetcode id=654 lang=golang
 *
 * [654] Maximum Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func constructMaximumBinaryTree(nums []int) *TreeNode {
	var buildTree func(nums []int, leftIndex, rightIndex int) *TreeNode
	buildTree = func(nums []int, leftIndex, rightIndex int) *TreeNode {
		if rightIndex-leftIndex < 1 {
			return nil
		}
		if rightIndex-leftIndex == 1 {
			return &TreeNode{Val: nums[leftIndex]}
		}

		maxIndex := leftIndex
		maxValue := nums[leftIndex]
		for i := leftIndex + 1; i < rightIndex; i++ {
			if nums[i] > maxValue {
				maxValue = nums[i]
				maxIndex = i
			}
		}

		root := &TreeNode{Val: maxValue}
		root.Left = buildTree(nums, leftIndex, maxIndex)
		root.Right = buildTree(nums, maxIndex+1, rightIndex)
		return root
	}
	return buildTree(nums, 0, len(nums))
}

// @lc code=end

